service/history/handler.go

```diff
@@ -1624,11 +1625,28 @@ func (h *handlerImpl) GetReplicationMessages(
 
 	wg.Wait()
 
+	responseSize := 0
+	maxResponseSize := h.config.GRPCMaxSizeInByte()
+
 	messagesByShard := make(map[int32]*types.ReplicationMessages)
 	result.Range(func(key, value interface{}) bool {
 		shardID := key.(int32)
 		tasks := value.(*types.ReplicationMessages)
-		messagesByShard[shardID] = tasks
+
+		size := proto.FromReplicationMessages(tasks).Size()
+		if (responseSize + size) >= maxResponseSize {
+			// Log shards that did not fit for debugging purposes
```

Comment:
Would you mind adding a metric here too? 

I assume it's safe for a partial response to come back? These'll be picked up next iteration of the replication fetch?

Reply:
Yes it is safe, remaining bits will be returned with next request.

Reply:
IMHO such metric is not strictly necessary as we already log it. And for alerting this will be detected by new ly added large replication backlog alert.

Reply:
we can set alerts based on metrics, but not logs though

Reply:
ok, sounds good